x/observer/keeper/utils.go

```diff
@@ -147,3 +147,33 @@ func (k Keeper) CheckObserverSelfDelegation(ctx sdk.Context, accAddress string)
 	}
 	return nil
 }
+
+// VoteOnBallot finds a ballot or creates a new one if not found,
+// and casts a vote on it. Then proceed to check if the vote has been finalized.
+// This function holds generic logic for all types of votes.
+func (k Keeper) VoteOnBallot(
+	ctx sdk.Context,
+	chain chains.Chain,
+	ballotIndex string,
+	observationType types.ObservationType,
+	voter string,
+	voteType types.VoteType,
+) (
```

Comment:
Can we add a unit test for this function?

Reply:
I was working on it, added right now. :)